Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -288,87 +247,176 @@ public object Deserialize(IIonReader reader, Type type, IonType ionType)
                 {
-                    return new IonDecimalSerializer().Deserialize(reader);
+                    return this.primitiveSerializers[typeof(decimal)].Deserialize(reader);
                 }
-                return new IonBigDecimalSerializer().Deserialize(reader);
+                return this.primitiveSerializers[typeof(BigDecimal)].Deserialize(reader);
             }
 
             if (ionType == IonType.Blob) 
             {
                 if (reader.GetTypeAnnotations().Any(s => s.Equals(IonGuidSerializer.ANNOTATION))
                     || type.IsAssignableTo(typeof(Guid)))
                 {
-                    return new IonGuidSerializer(options).Deserialize(reader);
+                    return this.primitiveSerializers[typeof(Guid)].Deserialize(reader);
                 }
-                return new IonByteArraySerializer().Deserialize(reader);
+                return this.primitiveSerializers[typeof(byte[])].Deserialize(reader);
             }
 
             if (ionType == IonType.String) 
             {
-                return new IonStringSerializer().Deserialize(reader);
+                return this.primitiveSerializers[typeof(string)].Deserialize(reader);
             }
 
             if (ionType == IonType.Symbol) 
             {
-                return new IonSymbolSerializer().Deserialize(reader);
+                return this.primitiveSerializers[typeof(SymbolToken)].Deserialize(reader);
             }
 
             if (ionType == IonType.Timestamp) 
             {
-                return new IonDateTimeSerializer().Deserialize(reader);
+                return this.primitiveSerializers[typeof(DateTime)].Deserialize(reader);
             }
 
             if (ionType == IonType.Clob) 
             {
-                return new IonClobSerializer().Deserialize(reader);
+                return this.clobSerializer.Deserialize(reader);
             }
 
             if (ionType == IonType.List) 
             {
-                return NewIonListSerializer(type).Deserialize(reader);
+                this.listSerializer.SetListType(type);
+                return this.listSerializer.Deserialize(reader);
             }
 
             if (ionType == IonType.Struct) 
             {
-                return new IonObjectSerializer(this, options, type).Deserialize(reader);
+                this.objectSerializer.targetType = type;
+                return this.objectSerializer.Deserialize(reader);
             }
 
-            throw new NotSupportedException("Don't know how to Deserialize this Ion data. Last IonType was: " + ionType);
+            throw new NotSupportedException($"Data with Ion type {ionType} is not supported for deserialization");
         }
 
         public T Deserialize<T>(IIonReader reader)
         {
             return (T) Deserialize(reader, typeof(T));
         }
+        
+        private bool ValidateCustomSerializer(Type type, dynamic serializer)

```

Comment:
Why is this validation necessary?

Reply:
Unfortunately we can't have a dictionary of `Type` and a generic of that type. So instead we are using a dictionary of `Type` with a `dynamic` object.
Since it's `dynamic`, we need to verify that the user provided a valid `IonSerializer<T>` for the given type.

Reply:
If you're able to change the dictionary to be `IonSerializer<dynamic>` instead of just `dynamic` I think think this check becomes trivial, since you'd only need to validate that the type of the serializer is already present for the primitives.